Userland/Libraries/LibCrypto/BigInt/Algorithms/ModularPower.cpp

```diff
@@ -49,4 +49,239 @@ void UnsignedBigIntegerAlgorithms::destructive_modular_power_without_allocation(
+        result.resize_with_leading_zeros(num_words);
+        return;
+    }
+
+    // We have a carry, so we're "one bigger" than we need to be.
+    // Subtract the modulo from the result (the top half of z), and write it to the bottom half of Z since we have space.
+    // (With carry, of course.)
+    UnsignedBigInteger::Word c { 0 };
+    for (size_t i = 0; i < num_words; ++i) {
+        UnsignedBigInteger::Word z_digit = z.m_words[num_words + i];
+        UnsignedBigInteger::Word modulo_digit = modulo.m_words[i];
+        UnsignedBigInteger::Word new_z_digit = z_digit - modulo_digit - c;
+        z.m_words[i] = new_z_digit;
+        // Detect if the subtraction underflowed - from "Hacker's Delight"
+        c = ((modulo_digit & ~z_digit) | ((modulo_digit | ~z_digit) & new_z_digit)) >> (UnsignedBigInteger::BITS_IN_WORD - 1);
+    }
+
+    // Return the bottom num_words bytes of Z (with the carry bit handled)
+    z.m_words.resize(num_words);
+    result.set_to(z);
+    result.resize_with_leading_zeros(num_words);
+}
+
+/**
+ * Complexity: still O(N^3) with N the number of words in the largest word, but less complex than the classical mod power.
+ * Note: the montgomery multiplications requires an inverse modulo over 2^32, which is only defined for odd numbers.
+ */
+void UnsignedBigIntegerAlgorithms::montgomery_modular_power_with_minimal_allocations(
+    UnsignedBigInteger const& base,
+    UnsignedBigInteger const& exponent,
+    UnsignedBigInteger const& modulo,
+    UnsignedBigInteger& temp_z,
+    UnsignedBigInteger& rr,
+    UnsignedBigInteger& one,
+    UnsignedBigInteger& z,
+    UnsignedBigInteger& zz,
+    UnsignedBigInteger& x,
+    UnsignedBigInteger& temp_for_swap,
+    UnsignedBigInteger& output)
+{
+    VERIFY(modulo.is_odd());
+
+    // Note: While this is a constexpr variable for clarity and could be changed in theory,
+    // various optimized parts of the algorithm rely on this value being exactly 4.
+    constexpr size_t window_size = 4;
+
+    size_t num_words = modulo.trimmed_length();
+    UnsignedBigInteger::Word k = inverse_wrapped(modulo.m_words[0]);
+
+    one.set_to(1);
+
+    // rr = ( 2 ^ (2 * modulo.length() * BITS_IN_WORD) ) % modulo
+    shift_left_by_n_words(one, 2 * num_words, x);
+    divide_without_allocation(x, modulo, temp_z, one, z, zz, temp_for_swap, rr);
+    rr.resize_with_leading_zeros(num_words);
+
+    // x = base [% modulo, if x doesn't already fit in modulo's words]
+    x.set_to(base);
+    if (x.trimmed_length() > num_words)
+        divide_without_allocation(base, modulo, temp_z, one, z, zz, temp_for_swap, x);
+    x.resize_with_leading_zeros(num_words);
+
+    one.set_to(1);
+    one.resize_with_leading_zeros(num_words);
+
+    // Compute the montgomery powers from 0 to 2^window_size. powers[i] = x^i
+    UnsignedBigInteger powers[1 << window_size];
+    almost_montgomery_multiplication_without_allocation(one, rr, modulo, temp_z, k, num_words, powers[0]);
```

Comment:
This table is pretty chunky (almost 32K), maybe we should re-evaluate whether Bigints need 512 words of inline storage...
On that note, I think I remember Vector never going back to its inline storage after the size overflows it, if that's still the case, should probably make that a little smarter (not necessarily as part of this PR ofc)

Reply:
I tried dropping it down to several values :
- 128 words didn't have any measurable performance impact. The average run time for 10 runs was at 0.63s, which is about the same given the high variance.
- 32 words (the limit to prevent unneeded allocations in montgomery powers) has around the same average as well (0.61s).
- 8 words (the absolute minimum to store a 256-bit value) stayed around 0.63s as well
- 1 word (too small to store anything of value) has the slightest of lower average performance (around 0.67s) with a higher maximum (0.90s), but that still largely fits inside of the variance.
The impact on the RSA test specifically is exactly none (0.18s average with or without the changes). So I don't think this value has an impact at all on the performances.

Since this doesn't have an impact, what value do you think is best to drop it down to?

Note: when I dropped this value too low (to e.g. 64 words), the tests hit a VERIFY on an `unchecked_append` in the `set_bit_inplace` method - this is an actual issue that woudl have bit us later. I added a new commit at the top of the branch that fixes this :^)

Reply:
I think the main problem is that `Vector` uses the inline capacity until it can't, then it will _never_ go back to using it unless `clear()` is called, which can only happen when called explicitly, or with `shrink(0)`.
So I think keeping a smallish inline buffer in case the values are small enough (e.g. 32) is good enough.

But I think we should deal with `Vector` not going back to inline capacity, maybe `clear_with_capacity()` should switch to the inline storage, but keep the outline buffer around for when it runs out of inline storage...

Reply:
Ok, I switched the default inline size to 32, and the rest can happen in another, wider-reaching PR !

Reply:
Actually, there's a big performance impact when below 64 words when running the tests inside of Serenity (instead of via Lagom).
The (single-run) results are :
- 256 words => ~1.6s
- 128 words =>  ~1.7s
- 64 words => ~1.5s
- 32 words => ~3.6s

I will change it to 64 words instead of 32 words, which will still give us the stack size gains without affecting performance.